### PR TITLE
Download files via fileById

### PR DESCRIPTION
**Pull request by Puepis**
## Notion ticket link

https://www.notion.so/uwblueprintexecs/Download-files-via-fileById-cce5b4acf4b0455981e6452e294eec77

## Implementation description

- implement `fileById` query by converting file data to a base64 string and returning it in a `DownloadFileDTO`

## Steps to test

1. start docker containers
2. in any directory, run `curl http://localhost:5000/graphql -F operations='{"query": "mutation ($file: Upload!) { createFile(file: $file) { ok }}", "variables": {"file": null }}' -F map='{ "0": ["variables.file"]}' -F 0=@file_name.txt`, replacing `file_name.txt` with any file in the directory to upload the file
3. in Altair, login as any user and run the query below
4. verify that the query succeeds and the file string is returned
5. if you want to verify that the string encoded properly, copy and run the following code in `codesandbox.io`

```
// Retrieved from https://developer.mozilla.org/en-US/docs/Glossary/Base64
function b64_to_utf8(str) {
  return decodeURIComponent(escape(atob(str)));
}
const b64String = "FILE_STRING_HERE";
const decoded = b64_to_utf8(b64String);
console.log(decoded);
```
**Note**: I'm not sure if the above ^ works for decoding non-`.txt` files

#### fileById query
```
query Download {
  fileById(id: 1) {
    id
    file
  }
}
```

## What should reviewers focus on?

- does it work for files that aren't `.txt` files? e.g. `.pdf`, `.docx`

Also, should we include a `fileName` field in the `DownloadFileDTO`? I'm thinking it'll be useful to at least return the file extension

## Checklist

- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] For backend changes, I have run the appropriate linters: `docker exec -it planet-read_py-backend_1 /bin/bash -c "black . && isort --profile black ."` and I have generated new migrations: `flask db migrate -m "<your message>"`
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Review by gaoxk — COMMENTED** [review]
works as outlined!

Question: Imagine our frontend makes this gql request. With the decoded information, how would the frontend save the file? Would the frontend need to know what kind of file (eg txt vs pdf) it is receiving?

**Comment by Puepis** [comment]
> With the decoded information, how would the frontend save the file? 

I did some searching and it seems like we would need to convert the base64 string to a blob, then either use a library like [file-saver](https://www.npmjs.com/package/file-saver) or use `URL.createObjectURL(new Blob([blob]))` to create a link and inject it into the DOM

Some links: 
- https://javascript.plainenglish.io/how-to-create-download-and-upload-files-in-react-apps-80893da4247a
- https://stackoverflow.com/a/63965930

> Would the frontend need to know what kind of file (eg txt vs pdf) it is receiving?

yup I believe so, I can add a field called `ext` to the DTO to return the file extension

File: backend/python/app/services/implementations/file_service.py [hunk 1]
```diff
@@ -22,7 +22,10 @@ def get_file(self, id):
             # encode file data as base64 string
             file_bytes = open(file.path, "rb").read()
             encoded = b64encode(file_bytes).decode("utf-8")
-            return DownloadFileDTO(id=id, file=encoded)
+
+            # extract extension and path
```
Comment (gaoxk) [inline backend/python/app/services/implementations/file_service.py:26]: nit: drop the `.` before the extension
```
{
  "data": {
    "fileById": {
      "id": 1,
      "ext": ".txt"
    }
  }
}
```

**Review by gaoxk — APPROVED** [review]
noice

**Pull request merged**